### ISSUE TITLE
AddOpenIdConnect options.Authority overriden with current host behind proxy (UseForwardedHeaders)

### ISSUE THREAD

**Issue by Genata** [dotnet/aspnetcore#31383]
I host my application behind reverse proxy inside container. I'm using IdentityServer4 and I'm trying to make oidc SSO with Azure AD. It works when the application is not using app.UseForwardedHeaders(opts), but I need them.

.NET Core version: 3.1

Forwarded headers configurations:
```
var opts = new ForwardedHeadersOptions
    {
        ForwardedHeaders = ForwardedHeaders.XForwardedFor | ForwardedHeaders.XForwardedProto |
                           ForwardedHeaders.XForwardedHost
    };
opts.KnownProxies.Add(ipAddress);
app.UseForwardedHeaders(opts);
```

OpenID Connect Configurations:
```
services.AddAuthentication( options =>
{
    options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
    options.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
}
    )
    .AddOpenIdConnect("aadrm", "Azure AD", options =>
    {
        options.SignInScheme = IdentityServerConstants.ExternalCookieAuthenticationScheme;
        options.SignOutScheme = IdentityServerConstants.SignoutScheme;

        options.Authority = "https://login.microsoftonline.com/{tenantID...}/v2.0";
        
        options.ClientId = ".....";
        options.ClientSecret = @"........";
        options.ResponseType = OpenIdConnectResponseType.CodeIdToken; 
    });
```

My Controller Action:
```
[HttpGet]
public IActionResult ExternalLoginChallenge(string provider, string returnUrl)
{
    var callbackUrl = Url.Action(nameof(ExternalLoginCallback));

    var props = new AuthenticationProperties
    {
        RedirectUri = callbackUrl,
        Items =
        {
            { "scheme", provider },
            { "returnUrl", returnUrl }
        }
    };

    return Challenge(props, provider);
}
```

When I initiate the sign in process it should redirect me to the Microsoft's sign in page, but instead it redirects me back to my host with this url: https://mylocalapp.com:443/{tenantId}/oauth2/v2.0/authorize?client_id=...&redirect_uri=HTTPS%3A%2F%2Fmylocalapp.com%2Fsignin-oidc&response_type=code%20id_token&scope=openid%20profile&response_mode=form_post&nonce=........&state=............&x-client-SKU=ID_NETSTANDARD2_0&x-client-ver=6.8.0.0

When I remove app.UseForwardedHeaders(opts) it works normally (redirects me to MS login page). It seems that UseForwardedHeaders overrides the OpenIdConnect Authority Address host.

Can you help me, I don't understand why it redirects me back to my host?

Thanks.

**Comment by blowdart** [comment]
That's odd. Have you tried capturing a network trace using the browser f12 tools to see the flow when you attempt to login? Is your machine AAD joined by any change?

**Comment by Genata** [comment]
Yes, there are only 2 requests in the network trace. 
One for the ExternalLoginChallenge and it redirects to the https://mylocalapp.com/{tenantId}/oauth2/v2.0/authorize?client_id=...

You can see the file attached (I've removed some sensitive data from the file). 
[mylocalapp.com.har.txt](https://github.com/dotnet/aspnetcore/files/6229957/mylocalapp.com.har.txt)

My machine is part of our corporate AAD.

**Comment by blowdart** [comment]
This is super weird. At this point I don't know, I've not seen this in our code, but then we don't tend to configure identity server too. Given the redirect can be overridden in code, I'd suggesting asking them first if they've seen it.

**Comment by Genata** [comment]
Thank you for the help and the answer, I appreciate it!

I've found the issue, it was not caused by docker, or the .net application, it was caused by the IIS reverse proxy configurations, that I'm using locally - Application Request Routing -> Server Proxy Settings -> Reverse rewrite host in response headers. 

![image](https://user-images.githubusercontent.com/15254783/113126268-24dcf000-9220-11eb-9426-c43c16e38529.png)

I've found that when I revised the response headers returned by the application during the redirecr request. I should have done that before posting this issue. :/
